packages/app/obojobo-express/util/promise_provider.js

```diff
@@ -0,0 +1,45 @@
+// possible npm packages to use instead:
+// emittery
+// p-event
+const logger = oboRequire('logger')
+const promiseProviders = {}
+
+// Register promise to an event
+// your provider mus be an async function or Promise
+const registerPromiseProvider = (eventName, provider) => {
+	if (!promiseProviders[eventName]) promiseProviders[eventName] = new Set()
+	const providersForType = promiseProviders[eventName]
+	providersForType.add(provider)
+}
+
+// execute all promises for an eventName in order with the provided arugments
+// setting returnOnFirstNonNullResponse to true will return immediately
+// when a promise returns a non-null value, skipping the remaining items
+// always returns an array
+const callPromiseProviders = async (
+	eventName,
+	providerArgs,
+	returnOnFirstNonNullResponse = false
```

Comment:
Notes to self: perhaps this could be `returnIfTrueFn` passing something like `value => value === true` would return as soon as a `true` is found, but be flexible enough to handle other options.